### DISCUSSION TITLE
Streaming to Android

### DISCUSSION THREAD

**Discussion by rootgordon** [MusicPlayerDaemon/MPD#1656] [Q&A] [answered]
I'm running mpd on a Rock Pi 4, using Manjaro Arm. This works as well as I could ask, so now I'm trying to use this as the base for streaming my music to my phone. I started out with [these](https://www.joram.io/blog/android-streaming-mpd/) handy directions on how to make this work. I am using M.A.L.P. as the client on Android.

Now M.A.L.P. does see the database but when I want to play an album it seemingly just skips through to the last track, not playing a single second of sound. I've tried this with both http and nfs for the music_directory. Here's my mpd.conf on the phone:

`#music_directory "nfs://192.168.1.10/home/01music/`
`  music_directory "http://192.168.1.10:8000"`
`  bind_to_address "127.0.0.1"`
                                                                                  
`  database {`
`     plugin "proxy"`
`     host "192.168.1.10"`
`     port "6600"`
`  }`

Using the default config the client will both see and play the music that is stored on my phone, so I don't think the client is doing anything wrong here. But the server is also not giving me any unusual messages. I can access the directory via http using a browser and I can mount and read the nfs-directory from another PC so those shouldn't be the problem either... Is there some way to run the mpd-server in debug-mode on android to get to the bottom of this? Or am I overlooking some painfully obvious mistake I made here?

**Comment by MaxKellermann** [comment]
Check https://www.musicpd.org/help/ there you'll find generic troubleshooting instructions, most of which are also relevant for Android.

**Comment by rootgordon** [comment]
Thanks for the suggestion but I already checked that. I don't understand how I can pass those debug-options to mpd on android though, so here I am...

**Reply by MaxKellermann** [reply]
Let me copy'n'paste for you:
> Check the log file. Configure log_level "verbose" or pass --verbose to mpd.

**Reply by rootgordon** [reply]
Something painfully obvious then. I'm still not getting more output from mpd than I was before.

**Comment by skidoo23** [comment]
May I suggest an easy solution?

Define http streaming output in mpd.conf of central server (rootgordon's rockpi) - eg generic lame encoder with 320 kbit/s and port 8000 (YMMV)

```
audio_output {
        type            "httpd"
        tags            "yes"
        name            "httpd on port 8000"
        always_on       "yes"
        encoder         "lame"                  # no flac or vorbis
        port            "8000"
        bind_to_address "0.0.0.0"               # optional, IPv4 or IPv6
        bitrate         "320"                   # do not define if quality is defined
        format          "44100:16:2"
        max_clients     "0"                     # optional 0=no limit
}
```

On your Android device with MALP client

1. server properties / output / ear symbol / enable "httpd on port 8000" output
2. profile / edit profile / enable streaming from server
3. profile / edit profile / streaming URL http://IP of your rockpi:8000
4. save modified profile
5. main player screen / tap on the three dots / start streaming
6. wait five seconds
7. enjoy the music

**Reply by rootgordon** [reply]
Easy, yes. In fact I've had this setup before but I was hoping to get rid of the five second delay (which feels ridiculously long) by running a satellite server on the phone. Should've mentioned this in the OP.

**Reply by skidoo23** [reply]
Looks you are talking about satellite setup. Not quite streaming audio it is just playling remote sound files ;) Always wanted a similar setup with nfs. Please allow me some time for testing. Would you be so kind to supply us in the meantime with verbose MPD and nfs logfiles?

**Comment by skidoo23** [comment]
Here we go. Lets assume 192.168.1.0/24 network, 192.168.1.1 server with /srv/music directory shared via NFS.

On main server /etc/exports
`/srv/music     192.168.1.0/24(ro,no_subtree_check,insecure)`

On the Android device Android\data\org.musicpd\files\mpd.conf (UTF-8 encoded)
```
music_directory			"nfs://192.168.1.1/srv/music"
log_level			"verbose"
database {
				plugin	"proxy"
				host	"192.168.1.1"
				port	"6600"
}
```

Connect with Android client software (eg MALP) to localhost:6600. Enjoy.

**Comment by rootgordon** [comment, marked as answer]
So recently I've had to get a new phone, prompting me to give this another shot. Initially I ran into the same problem but through sheer force of questioning everything I finally found out what was wrong (at least in the case of an nfs-mount): The files on the server need to be world-readable.

It's kinda obvious in hindsight, I don't even know what user-ID mpd, or android for that matter, is using. I hate when these threads remain unresolved so this is my solution for anyone who might run into the same problem and finds this.

Also here's a one-liner I used to chmod all the files recursively (based on an answer in [this thread](https://superuser.com/questions/91935/how-to-recursively-chmod-all-directories-except-files)):

`chmod -R a=r-wx,u=wr,g=wr,a+X *`

It leaves directories world-executable while only setting the read-bit on files, you may want to adjust user and group-rights to your liking. It's handy for large collections, which anyone attempting this setup probably has.

Using this with wireguard for remote-access works like a charm and clears up loads of space on my phone so I can take more pictures of my cat. Amazing.

**Reply by MaxKellermann** [reply]
Simpler version: `chmod -R a+rX /music/directory`
